src/sieppari/async.cljc

```diff
@@ -21,20 +27,33 @@
    (extend-protocol AsyncContext
      Object
      (async? [_] false)
+     ; Given the implementation of enter/leave,
+     ; `continue` won't be called, and therefore,
+     ; the function call does not need to be bound
      (continue [t f] (f t))
      (await [t] t)))
 
 #?(:cljs
    (extend-protocol AsyncContext
      default
      (async? [_] false)
+     ; Given the implementation of enter/leave,
+     ; `continue` won't be called, and therefore,
+     ; the function call does not need to be bound
      (continue [t f] (f t))))
 
 #?(:clj
    (extend-protocol AsyncContext
      clojure.lang.IDeref
      (async? [_] true)
-     (continue [c f] (future (f @c)))
+     (continue [c f]
+       (cond
+         ; Make sure there is 
+         (map? @c)
+           (with-bindings (or (:bindings @c) {})
+             (future ((bound-fn* f) @c)))
+         :else
+           (future (f @c))))
```

Comment:
Not sure this is necessary. Should elide the `cond` and use the `with-bindings` form